pandas/core/arrays/categorical.py

```diff
@@ -1259,7 +1259,10 @@ def __array__(self, dtype=None):
             if dtype==None (default), the same dtype as
             categorical.categories.dtype
         """
-        ret = take_1d(self.categories.values, self._codes)
+        # Need asarray, in case self.categories.values is an ExtensionArray
+        # e.g. in a PeriodIndex. More generally, any Index backed by an EA.
+        values = np.asarray(self.categories.values)
+        ret = take_1d(values, self._codes)
```

Comment:
At the moment the cython code backing `take_1d` is partially-converted from requiring np.ndarray to more general typed memoryviews.  It might be worth looking into what it would take to make cython recognize EAs the way it does ndarrays.

Reply:
> make cython recognize EAs the way it does ndarrays.

I haven't written much cython, but that sounds difficult since EA places no restrictions on how the data are actually stored. But if the underlying storage does implement the buffer protocol, it'd be nice to be able to use that down in our algos... I'm just not sure whether the EA would implement the buffer protocol as well, or whether pandas would be responsible for extracting the storage array that does implement it.